internal/pipeline/executor.go

```diff
@@ -0,0 +1,271 @@
+	root *pkg
+
+	logger *zap.SugaredLogger
+
+	// TODO (droot): wire sink-mode here
+
+	// pkgs refers to the discovered pkgs where
+	// It's a map where key refers to package's absolute path and value
+	// refers to the pkg.
+	pkgs map[string]*pkg
+
+	// hydrationStates for each package (dry, hydrating or wet) in the hydration DAG
+	hydrationStates map[string]hydrationState
+}
+
+//
+// pkg represents a kpt package (node) in the hydration DAG.
+//
+// TODO(droot): kpt will certainly have pkg abstraction, may be embed that here to
+// avoid duplication.
+type pkg struct {
+	// absolute path on the filesystem
+	path string
+
+	// pipeline contained in the pkg
+	pipeline *Pipeline
+
+	// tracks the state of hydration (dry, hydrating, wet)
+	state hydrationState
+
+	// resources post hydration
+	resources []*yaml.RNode
+}
+
+// hydrationState represent hydration state of a pkg.
+type hydrationState int
+
+// constants for all the hydration states
+const (
+	Dry hydrationState = iota
+	Hydrating
+	Wet
+)
+
+func (s hydrationState) String() string {
+	return []string{"Dry", "Hydrating", "Wet"}[s]
+}
+
+//
+// hydrate hydrates a pkg and returns wet resources.
+// TODO(droot): makes these comments really useful :)
+func hydrate(p *pkg, hctx *hydrationContext) (resources []*yaml.RNode, err error) {
+	logger := hctx.logger
+
+	state, found := hctx.hydrationStates[p.path]
+	if found {
+		switch state {
+		case Hydrating:
+			// we have a cycle
+			// TODO(droot): improve the error to contain useful information
+			err = fmt.Errorf("found cycle in dependencies for pkg %s", p.path)
+			return
+		case Wet:
+			// already hydrated, we sort of assume pkgs in hctx will be populated but we
+			// can be defensive here
+			resources = hctx.pkgs[p.path].resources
+			return
+		default:
+			err = fmt.Errorf("invalid pkg state %v detected", state)
+			return
+		}
+	}
+	// mark the pkg in hydrating
+	hctx.hydrationStates[p.path] = Hydrating
+	// add to the graph
+	hctx.pkgs[p.path] = p
+
+	// read the pkg, the pipeline, dependencies etc.
+	err = readPkg(p)
+	if err != nil {
+		err = fmt.Errorf("failed to read the pkg %w", err)
+		return
+	}
+
+	// input resources for current pkg's hydration
+	var input []*yaml.RNode
+	includeLocalResources := false
+	// here we explore the p.sources to determine sub packages etc.
+	for _, s := range expand(p.pipeline.Sources, p.path) {
+		// TODO(droot): sync with pipeline's defaults once Donny's PR is merged
+		if s == "." {
+			// current pkg
+			// will be hydrated as we drop of this loop
+			includeLocalResources = true
+		} else {
+			// TODO(droot): recursive hydration is broken for now, needs more work
+			var depResources []*yaml.RNode
+			depPkg := &pkg{
+				path: s,
+			}
+			depResources, err = hydrate(depPkg, hctx)
+			if err != nil {
+				err = fmt.Errorf("failed to hydrated sub pkg %s %w", s, err)
+				return
+			}
+			input = append(input, depResources...)
+		}
+	}
+
+	if includeLocalResources {
```

Comment:
We should extract out such common operations into functions or methods.

If going with OOP approach:

p.readLocalResources()